File: sample-push-notifications/sample-push-notifications/PushViewController.m
//
//  ViewController.m
//  sample-push-notifications
//
//  Created by Injoit on 6/11/15.
//  Copyright (c) 2015 QuickBlox. All rights reserved.
//

#import "PushViewController.h"
#import <Quickblox/Quickblox.h>
#import <SVProgressHUD.h>
#import "SAMTextView.h"
#import "UIViewController+InfoScreen.h"
#import "Profile.h"
#import "NotificationsProvider.h"
#import "AppDelegate.h"
#import "RootParentVC.h"

@interface PushViewController () <UITableViewDataSource, UITableViewDelegate, NotificationsProviderDelegate>

#pragma mark - Properties
@property (weak, nonatomic) IBOutlet SAMTextView *pushMessageTextView;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *sendPushButton;
@property (weak, nonatomic) IBOutlet UISegmentedControl *segmentedControl;

@property (assign, nonatomic) PushType currentPushType;
@property (nonatomic, strong) NSMutableArray *pushMessages;
@property (nonatomic, strong) NotificationsProvider *notificationsProvider;

@end

@implementation PushViewController

#pragma mark - Life Cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.notificationsProvider = [[NotificationsProvider alloc] init];
    self.notificationsProvider.delegate = self;
    self.pushMessages = [NSMutableArray array];
    self.currentPushType = PushTypeAPNS;
    
    NSDictionary* attributes = @{NSFontAttributeName : [UIFont systemFontOfSize:17.0f],
                                 NSForegroundColorAttributeName : [UIColor colorWithWhite:0.0f alpha:0.3f]};
    self.pushMessageTextView.attributedPlaceholder = [[NSAttributedString alloc] initWithString:@"Enter push message here" attributes:attributes];
    self.pushMessageTextView.textContainerInset = (UIEdgeInsets){10.0f, 10.0f, 0.0f, 0.0f};
    
    CALayer *bottomBorder = [CALayer layer];
    bottomBorder.frame = CGRectMake(0.0f, self.pushMessageTextView.frame.size.height - 1.0f, self.pushMessageTextView.frame.size.width, 1.0f);
    bottomBorder.backgroundColor = [UIColor colorWithRed:200.0f/255.0f
                                                   green:199.0f/255.0f
                                                    blue:204.0f/255.0f
                                                   alpha:1.0f].CGColor;
    [self.pushMessageTextView.layer addSublayer:bottomBorder];
    
    self.sendPushButton.enabled = NO;
    
    Profile *profile = [[Profile alloc] init];
    if (profile.isFull) {
        self.title = profile.fullName;
        self.sendPushButton.enabled = YES;
    } else {
        [SVProgressHUD showErrorWithStatus:@"You are not authorized."];
    }
    
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Logout"
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(logoutButtonPressed:)];
    [self addInfoButton];
}

#pragma mark - Actions
- (IBAction)pushTypeDidChanged:(UISegmentedControl *)sender {
    if (sender.selectedSegmentIndex == 0) {
        self.currentPushType = PushTypeAPNS;
    } else if (sender.selectedSegmentIndex == 1) {
        self.currentPushType = PushTypeAPNSVOIP;
    }
}

- (void)sendPushWithMessage:(NSString *)message {
    
    Profile *profile = [[Profile alloc] init];
    if (!profile.isFull) {
        return;
    }
    
    NSString *currentUserId = [NSString stringWithFormat:@"%@", @(profile.ID)];
    
    if (self.currentPushType == PushTypeAPNS) {
        [SVProgressHUD showWithStatus:@"Sending a APNS Push"];
        
        [QBRequest sendPushWithText:message toUsers:currentUserId successBlock:^(QBResponse *response, NSArray *events) {
            [SVProgressHUD showSuccessWithStatus:@"Your message successfully sent"];
        } errorBlock:^(QBError *error) {
            [SVProgressHUD showErrorWithStatus:error.description];
        }];
        
    } else if (self.currentPushType == PushTypeAPNSVOIP) {
        [SVProgressHUD showWithStatus:@"Sending a VOIP Push"];
        NSDictionary *payload = @{
            @"message"  : message,
            @"ios_voip" : @"1",
            @"VOIPCall"  : @"1",
            @"alertMessage": message
        };
        NSData *data =
        [NSJSONSerialization dataWithJSONObject:payload
                                        options:NSJSONWritingPrettyPrinted
                                          error:nil];
        NSString *eventMessage =
        [[NSString alloc] initWithData:data
                              encoding:NSUTF8StringEncoding];
        
        QBMEvent *event = [QBMEvent event];
        event.notificationType = QBMNotificationTypePush;
        event.usersIDs = currentUserId;
        event.type = QBMEventTypeOneShot;
        event.message = eventMessage;
        
        [QBRequest createEvent:event
                  successBlock:^(QBResponse *response, NSArray<QBMEvent *> *events) {
            [SVProgressHUD showSuccessWithStatus:@"Your message successfully sent"];
        } errorBlock:^(QBResponse * _Nonnull response) {
            [SVProgressHUD showErrorWithStatus:response.description];
        }];
    }
}

#pragma mark Logout
- (void)logoutButtonPressed:(UIButton *)sender {
    [SVProgressHUD showWithStatus:@"Logout..."];
    
#if TARGET_OS_SIMULATOR
    [self logOut];
#else
    NSString *deviceIdentifier = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
    __weak __typeof(self)weakSelf = self;
    [QBRequest subscriptionsWithSuccessBlock:^(QBResponse * _Nonnull response, NSArray<QBMSubscription *> * _Nullable objects) {
        __typeof(weakSelf)strongSelf = weakSelf;
        dispatch_group_t deleteSubscriptionsGroup = dispatch_group_create();
        for (QBMSubscription *subscription in objects) {
            if ([subscription.deviceUDID isEqualToString:deviceIdentifier]) {
                dispatch_group_enter(deleteSubscriptionsGroup);
                [QBRequest deleteSubscriptionWithID:subscription.ID successBlock:^(QBResponse * _Nonnull response) {
                    dispatch_group_leave(deleteSubscriptionsGroup);
                    NSLog(@"[%@] Unregister Subscription request - Success",  NSStringFromClass([NotificationsProvider class]));
                } errorBlock:^(QBResponse * _Nonnull response) {
                    dispatch_group_leave(deleteSubscriptionsGroup);
                    NSLog(@"[%@] Unregister Subscription request - Error",  NSStringFromClass([NotificationsProvider class]));
                }];
            }
        }
        dispatch_group_notify(deleteSubscriptionsGroup, dispatch_get_main_queue(), ^{
            [strongSelf logOut];
        });
        
    } errorBlock:^(QBResponse * _Nonnull response) {
        if (response.status == 404) {
            [self showLoginScreen];
        }
    }];
#endif
}

- (void)showLoginScreen {
    [Profile clear];
    [self.rootParentVC showLoginScreen];
    [SVProgressHUD showSuccessWithStatus:@"Completed"];
}

- (void)logOut {
    __weak __typeof(self)weakSelf = self;
    [QBRequest logOutWithSuccessBlock:^(QBResponse * _Nonnull response) {
        __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf showLoginScreen];
    } errorBlock:^(QBResponse * _Nonnull response) {
        if (response.error.error) {
            [SVProgressHUD showErrorWithStatus:response.error.error.localizedDescription];
            return;
        }
    }];
}

- (IBAction)sendPush:(id)sender {
    [self.view endEditing:YES];
    NSString *message = self.pushMessageTextView.text;
    // empty text
    if([message length] == 0) {
        [SVProgressHUD showInfoWithStatus:@"Please enter some text"];
        return;
    }
    [self sendPushWithMessage:message];
    [self.pushMessageTextView resignFirstResponder];
    self.pushMessageTextView.text = nil;
}

#pragma mark - Internal Methods
- (void)didReceivePush:(NSArray<NSString *> *)messages {
    for (NSString *message in messages) {
        [self.pushMessages insertObject:message atIndex:0];
    }
    [self.tableView reloadData];
}

#pragma mark TableViewDataSource & TableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.pushMessages count];
}

- (UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"PushMessageCellIdentifier"];
    cell.textLabel.text = self.pushMessages[indexPath.row];
    
    return cell;
}

#pragma mark NotificationsProviderDelegate
- (void)notificationsProvider:(NotificationsProvider *)notificationsProvider didReceiveMessages:(NSArray<NSString *> *)messages {
    [self didReceivePush:messages];
}

- (void)notificationsProvider:(NotificationsProvider *)notificationsProvider willPresentMessage:(NSString *)message {
    [self didReceivePush:@[message]];
}

- (void)notificationsProvider:(NotificationsProvider *)notificationsProvider didReceiveIncomingVOIPPushWithMessage:(NSString *)message {
    [self didReceivePush:@[message]];
}

#pragma mark - RootParentVC
- (RootParentVC*)rootParentVC {
    return (RootParentVC *)[[UIApplication sharedApplication] delegate].window.rootViewController;
}

@end

File: sample-push-notifications-swift/sample-push-notifications-swift/Components/PushesScreen/PushViewController.swift
//
//  ViewController.swift
//  sample-push-notifications-swift
//
//  Created by Injoit on 3/19/19.
//  Copyright © 2019 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import SVProgressHUD

class PushViewController: UIViewController {
    
    //MARK: - Properties
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var sendPushButton: UIBarButtonItem!
    @IBOutlet weak var pushMessageTextView: PushTextView!
    @IBOutlet weak var segmentedController: UISegmentedControl!
    
    var pushMessages: [String] = []
    private var currentPushType: PushType = .apns
    
    lazy private var notificationsProvider: NotificationsProvider = {
        let notificationsProvider = NotificationsProvider()
        return notificationsProvider
    }()

    
    //MARK: - Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()

        notificationsProvider.delegate = self
        pushMessageTextView.placeholder = "Enter push message here"
        pushMessageTextView.textContainerInset = UIEdgeInsets(top: 10.0, left: 10.0, bottom: 0.0, right: 0.0)
        
        let bottomBorder = CALayer()
        bottomBorder.frame = CGRect(x: 0.0,
                                    y: pushMessageTextView.frame.size.height - 1.0,
                                    width: view.frame.size.width,
                                    height: 1.0)
        bottomBorder.backgroundColor = UIColor(red: 200.0/255.0,
                                               green: 199.0/255.0,
                                               blue: 204.0/255.0,
                                               alpha: 1.0).cgColor
        pushMessageTextView.layer.addSublayer(bottomBorder)

        sendPushButton.isEnabled = false
        
        let profile = Profile()
        if profile.isFull == true {
            sendPushButton.isEnabled = true
            title = profile.fullName
        } else {
            SVProgressHUD.showError(withStatus: "You are not authorized.")
        }
        
        let logoutButton = UIBarButtonItem(title:"Logout",
                                           style: .plain,
                                           target: self,
                                           action: #selector(didTapLogout(_:)))
        navigationItem.leftBarButtonItem = logoutButton
        
        addInfoButton()
    }
    
    //MARK: - Actions
    @IBAction func pushTypeDidChanged(_ sender: UISegmentedControl) {
        if sender.selectedSegmentIndex == 0 {
            currentPushType = .apns
        } else if sender.selectedSegmentIndex == 1 {
            currentPushType = .apnsVoip
        }
    }
    
    
    @objc private func didTapLogout(_ sender: UIBarButtonItem) {
        SVProgressHUD.show(withStatus: "Logouting...")
        SVProgressHUD.setDefaultMaskType(.clear)
        
        guard let uuidString = UIDevice.current.identifierForVendor?.uuidString else {
            return
        }
        #if targetEnvironment(simulator)
        logOut()
        #else
        
        QBRequest.subscriptions(successBlock: { [weak self] (response, subscriptions) in
            guard let subscriptions = subscriptions, subscriptions.isEmpty == false else {
                self?.logOut()
                return
            }
            let deleteSubscriptionsGroup = DispatchGroup()
            for subscription in subscriptions {
                if let subscriptionsUIUD = subscription.deviceUDID,
                   subscriptionsUIUD == uuidString {
                    deleteSubscriptionsGroup.enter()
                    QBRequest.deleteSubscription(withID: subscription.id) { (response) in
                        deleteSubscriptionsGroup.leave()
                        debugPrint("[NotificationsProvider] Unregister Subscription request - Success")
                    } errorBlock: { (response) in
                        deleteSubscriptionsGroup.leave()
                        debugPrint("[NotificationsProvider] Unregister Subscription request - Error")
                    }
                }
            }
            deleteSubscriptionsGroup.notify(queue: DispatchQueue.main) {
                self?.logOut()
            }
        }) { (response) in
            if response.status.rawValue == 404 {
                self.showLoginScreen()
            }
        }
        #endif
    }
    
    //MARK: - Internal Methods
    //MARK: - logOut flow
    fileprivate func showLoginScreen() {
        Profile.clear()
        AppDelegate.shared.rootViewController.showLoginScreen()
        SVProgressHUD.showSuccess(withStatus: "Complited")
    }
    
    private func logOut() {
        QBRequest.logOut(successBlock: { [weak self] response in
            self?.showLoginScreen()
        }) { response in
            debugPrint("[DialogsViewController] logOut error: \(response)")
        }
    }
    
    //MARK: - IActions
    @IBAction func sendPush(_ sender: UIBarButtonItem) {
        
        view.endEditing(true)
        guard let message = pushMessageTextView.text else {
            return
        }
        // empty text
        if message.isEmpty == true {
            SVProgressHUD.showInfo(withStatus: "Please enter some text")
        } else {
            sendPushWith(message)
            pushMessageTextView.resignFirstResponder()
            pushMessageTextView.text = nil;
        }
    }
    
    //MARK: - Internal Methods
    private func sendPushWith(_ message: String) {
        
        let profile = Profile()
        if profile.isFull == false {
            return
        }
        let currentUserId = "\(profile.ID)"

        switch currentPushType {
        case .apns:
            SVProgressHUD.show(withStatus: "Sending a APNS Push")
            QBRequest.sendPush(withText: message, toUsers: currentUserId, successBlock: { (response, events) in
                SVProgressHUD.showSuccess(withStatus: "Your message successfully sent")
                
            }) { (error) in
                SVProgressHUD.showError(withStatus: error.description)
            }
        case .apnsVoip:
            SVProgressHUD.show(withStatus: "Sending a VOIP Push")
            let payload = ["message": message,
                "ios_voip": "1",
                "VOIPCall": "1",
                "alertMessage": message
            ]
            let data = try? JSONSerialization.data(withJSONObject: payload,
                                                   options: .prettyPrinted)
            var eventMessage = ""
            if let data = data {
                eventMessage = String(data: data, encoding: .utf8) ?? ""
            }
            let event = QBMEvent()
            event.notificationType = QBMNotificationType.push
            event.usersIDs = currentUserId
            event.type = QBMEventType.oneShot
            event.message = eventMessage
            QBRequest.createEvent(event, successBlock: { response, events in
                SVProgressHUD.showSuccess(withStatus: "Your message successfully sent")
            }, errorBlock: { error in
                SVProgressHUD.showError(withStatus: error.description)
            })
        }
    }
    
    func didReceivePush(_ messages: [String]) {
        for message in messages {
            pushMessages.insert(message, at: 0)
        }
        tableView.reloadData()
    }
}

//MARK: - TableViewDataSource & TableViewDelegate
extension PushViewController: UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return pushMessages.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: "PushMessageCellIdentifier") else {
            return UITableViewCell()
        }
        cell.textLabel?.text = pushMessages[indexPath.row]
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
    }
}

extension PushViewController: NotificationsProviderDelegate {
    func notificationsProvider(_ notificationsProvider: NotificationsProvider, didReceive messages: [String]) {
        didReceivePush(messages)
    }
}
